crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -137,23 +137,33 @@ impl<'a> Checker<'a> {
     }
 }
 
+/// Visit an body of [`Stmt`] nodes within a type-checking block.
+macro_rules! visit_type_checking_block {
+    ($self:ident, $body:expr) => {{
+        let flags = $self.ctx.flags;
+        $self.ctx.flags |= ContextFlags::TYPE_CHECKING_BLOCK;
```

Comment:
Note: we're exposing the flags to enable the visitor to modify them, rather than exposing a bunch of write methods, which felt tedious to me. Open to input though.

Reply:
What's the reason for this being a macro vs a regular method or standalone function that takes `ctx` as an argument?

```rust
impl<'_> Checker<'_> {
    fn visit_type_checking_block(&mut self, body: &[Stmt]) {
        let flags = self.ctx.flags;
        self.ctx.flags |= ContextFlags::TYPE_CHECKING_BLOCK;
        self.visit_body(body);
        self.ctx.flags = flags;
    }
}
```

The same applies for the other macros (`visit_type_definition`...)

Reply:
I honestly can't remember, I assume there's some kind of Chesterton's Fence thing here but I will revisit it in a follow-up PR.

Reply:
A what 😅?

Reply:
Hahaha. You come across a fence, and you're not sure why it's there. But you probably don't want to tear it down without understanding why it exists :D